ext/storage_sources/gcp_store/gcp_storage_source.cpp

```diff
@@ -48,42 +57,162 @@ static int gcp_remove(WT_FILE_SYSTEM *, WT_SESSION *, const char *, uint32_t)
+    gcp_file_system *fs;
+    if ((fs = (gcp_file_system *)calloc(1, sizeof(gcp_file_system))) == nullptr) {
+        std::cerr << "gcp_customize_fs: unable to allocate memory for file system." << std::endl;
+        return ENOMEM;
+    }
+
+    // Set variables specific to GCP.
+    fs->storage_source = gcp;
+    fs->wt_file_system = wt_file_system;
+    fs->home_dir = home_dir;
+
+    // Create connection to google cloud.
+    try {
+        fs->gcp_conn = std::make_unique<gcp_connection>(bucket, obj_prefix);
+    } catch (std::invalid_argument &e) {
+        std::cerr << "gcp_customize_file_system: " << e.what() << std::endl;
+        return EINVAL;
+    }
+
+    // Map google cloud functions to the file system.
+    fs->file_system.fs_directory_list = gcp_object_list;
+    fs->file_system.fs_directory_list_single = gcp_object_list_single;
+    fs->file_system.fs_directory_list_free = gcp_object_list_free;
+    fs->file_system.terminate = gcp_file_system_terminate;
+    fs->file_system.fs_exist = gcp_file_exists;
+    fs->file_system.fs_open_file = gcp_file_open;
+    fs->file_system.fs_remove = gcp_remove;
+    fs->file_system.fs_rename = gcp_rename;
+    fs->file_system.fs_size = gcp_file_size;
+
+    // Add to the list of the active file systems. Lock will be freed when the scope is exited.
+    {
+        std::lock_guard<std::mutex> lock_guard(gcp->fs_list_mutex);
+        gcp->fs_list.push_back(fs);
+    }
+
+    *file_system = &fs->file_system;
 
     return 0;
 }
 
 static int
-gcp_add_reference(WT_STORAGE_SOURCE *store)
+gcp_add_reference(WT_STORAGE_SOURCE *storage_source)
 {
-    WT_UNUSED(store);
+    gcp_store *gcp = reinterpret_cast<gcp_store *>(storage_source);
+
+    if (gcp->reference_count == 0 || gcp->reference_count + 1 == 0) {
+        std::cerr << "gcp_add_reference: missing reference or overflow." << std::endl;
+        return EINVAL;
+    }
+
+    ++gcp->reference_count;
+
+    return 0;
+}
+
+// File handle close.
+static int
+gcp_file_close(WT_FILE_HANDLE *file_handle, WT_SESSION *session)
+{
+    WT_UNUSED(file_handle);
+    WT_UNUSED(session);
 
     return 0;
 }
 
 static int
 gcp_file_system_terminate(WT_FILE_SYSTEM *file_system, WT_SESSION *session)
 {
-    WT_UNUSED(file_system);
+    gcp_file_system *fs = reinterpret_cast<gcp_file_system *>(file_system);
+    gcp_store *gcp = reinterpret_cast<gcp_file_system *>(fs)->storage_source;
+
     WT_UNUSED(session);
 
+    /*
+     * It is currently unclear at the moment what the multi-threading will look like in the
+     * extension. The current implementation is NOT thread-safe, and needs to be addressed in the
+     * future, as multiple threads could call terminate leading to a race condition.
+     */
+    while (!fs->fh_list.empty()) {
+        gcp_file_handle *fh = fs->fh_list.front();
+        gcp_file_close((WT_FILE_HANDLE *)fh, session);
+    }
+
+    // Remove from the active filesystems list. The lock will be freed when the scope is exited.
```

Comment:
Update the comment to say what we are removing from the active filesystems list.